ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -31,29 +33,69 @@ def __init__(self):
         rospy.Subscriber('/current_pose', PoseStamped, self.pose_cb)
         rospy.Subscriber('/base_waypoints', Lane, self.waypoints_cb)
 
-        # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint below
-
+        # Add a subscriber for /traffic_waypoint and /obstacle_waypoint below
+        rospy.Subscriber('/traffic_waypoint', Int32, self.traffic_cb)
+        rospy.Subscriber('/obstacle_waypoint', Int32, self.obstacle_cb)
 
         self.final_waypoints_pub = rospy.Publisher('final_waypoints', Lane, queue_size=1)
 
-        # TODO: Add other member variables you need below
+        # Add other member variables you need below
+        self.base_lane = None
+        self.traffic_wp = None
+        self.obstacle_wp = None
 
         rospy.spin()
 
     def pose_cb(self, msg):
-        # TODO: Implement
+        #rospy.loginfo("pose_cb timestamp %s x=%d y=%d z=%d", msg.header.stamp, msg.pose.position.x,
+        #              msg.pose.position.y, msg.pose.position.z)
+        if self.base_lane != None:
+            quaternion = (msg.pose.orientation.x, msg.pose.orientation.y,
+                          msg.pose.orientation.z, msg.pose.orientation.w)
+            roll, pitch, yaw = tf.transformations.euler_from_quaternion(quaternion)
+            #rospy.loginfo("roll %f pitch %f yaw %f", roll, pitch, yaw)
+            
+            # the course hits an inflection point at x=2339 yaw=90+ at which point x starts to decrease
+            # next inflection at x=155 yaw=270+ at which point x starts to increase
+            # in front of the car = increasing x for yaw between 270 and 90 and decreasing x from 90 to 270
+            pub_waypoints = []
+            wp_cnt = 0
+            fwd = True
+            if yaw > 90.0 and yaw <= 270.0:  # exactly 90/270 handling is a bit arbitrary
+                fwd = False
+            for wp in self.base_lane.waypoints:
+                if ((fwd and wp.pose.pose.position.x > msg.pose.position.x) or
```

Comment:
I'm not too clear on this conditional—pose is in global coordinates, so won't we need to look at both x and y to see which waypoint is closest to the car?

Reply:
I think the car's primary motion is along the x axis so that defines what is in front.  (I remember reading something in the lesson about using x but I'd have to dig it up).  The conditional is due to wraparound in the waypoints.  There is a second file that has yaw in it and after digging through that for a while it seems consistent with the idea that 0 degrees faces forward and x increments in front of the car within 90 degrees of that.  On the other side x decrements in front of the car.